compiler/optimizer/OMRCFGSimplifier.cpp

```diff
@@ -333,6 +333,198 @@ bool OMR::CFGSimplifier::simplifyInstanceOfTestToCheckcast(bool needToDuplicateT
+            secondThrowBlock = isSecondCompareLT ? nextCompareBlock->getNextBlock() : secondCompareNode->getBranchDestination()->getNode()->getBlock();
+            retBlock = isSecondCompareLT ? secondCompareNode->getBranchDestination()->getNode()->getBlock() : nextCompareBlock->getNextBlock();
+            isLastBlockRetBlock = isSecondCompareLT;
+            secondChildBndChk = secondCompareNode->getSecondChild();
+            secondNodeRequiresDuplicate = true;
+         }
+      }
+   }
+
+   if (!patternMatched || !firstChildBndChk || !secondChildBndChk || firstChildBndChk == secondChildBndChk || throwBlock != secondThrowBlock) {
+      if (trace())
+         traceMsg(comp(), "Abort simplifyBoundCheckWithThrowException : %s\n", !patternMatched ? "Pattern did not matched" : 
+            (!firstChildBndChk ? "None of the ificmp matches (param < 0) or (param >= 0)" : 
+            (!secondChildBndChk ? "None of the ificmp matches (param < limit) or (param >= limit)" : 
+            ((firstChildBndChk == secondChildBndChk) ? "Both branch compares either (param,0) or (param,limit)" :
+            "Each of the branch jumps to different throw block"))));
+      return false;
+   }
+
+   if (trace())
+      traceMsg(comp(), "   Matched with second %s in block_%d\n", isSecondCompareLT ? "ificmplt" : "ificmpge",
+               nextCompareBlock->getNumber());
+   
+   TR::Node *throwNode = throwBlock->getLastRealTreeTop()->getNode();
+   if (throwNode->getNumChildren() < 1 || throwNode->getFirstChild()->getOpCodeValue() != TR::athrow) {
+      if (trace())
+         traceMsg(comp(), "Abort simplifyBoundCheckWithThrowException : %s\n", throwNode->getNumChildren() < 1 ? "Expected throwNode does not contain children" :
+         "The throwNode does not contain throw, pattern not matched");
+      return false;
+   }
+
+   if (!performTransformation(comp(), "%sReplace %s n%dn [%p] followed by a %s n%dn [%p] that throws (block_%d) with a BNDCHK to a catch which goes to block_%d\n", 
+            OPT_DETAILS, isCompareLT ? "ificmplt" : "ificmpge",
+            compareNode->getGlobalIndex(), compareNode, secondCompareNode->getGlobalIndex(), secondCompareNode, throwBlock->getNumber(), throwBlock->getNumber()))
+      return false;
+
+   _cfg->invalidateStructure();
+
+   TR::Block *compareBlock = _block;
+   if (hasExceptionPoint(compareBlock, treeTop)) 
+      compareBlock = compareBlock->split(treeTop, _cfg, true, false);
+
+   TR::Node *bndChkNode = TR::Node::createWithSymRef(TR::BNDCHK, 2, 2, 
+                              secondNodeRequiresDuplicate ? secondChildBndChk->duplicateTree(true) : secondChildBndChk,
+                              firstNodeRequiresDuplicate ? firstChildBndChk->duplicateTree(true) : firstChildBndChk,
+                              comp()->getSymRefTab()->findOrCreateArrayBoundsCheckSymbolRef(comp()->getMethodSymbol()));
+
+   bndChkNode = compareBlock->append(TR::TreeTop::create(comp(), bndChkNode))->getNode();
+
```

Comment:
I think this needs to be guarded by a `performTransformation` call.

Reply:
The earlier call to `performTransformation` is not good enough? Sorry, if this question sounds stupid, but I don't have the full picture yet :-) . According to my understanding, if transformation is not allowed, we better off with skipping the simplification process.

Reply:
Doh! Apologies, you are completely correct. I made a mistake missing this `performTransformation` call. Thanks for pointing that out.

Reply:
No worries. Also thank you for confirming my understanding, still learning :smiley: .